File: Lwjgl-Headless/src/main/java/cn/rwhps/lwjgl/headless/redirections/STBIImageRedirection.java
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package cn.rwhps.lwjgl.headless.redirections;

import cn.rwhps.lwjgl.headless.api.Redirection;
import cn.rwhps.lwjgl.headless.util.ByteBufferInputStream;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;

public enum STBIImageRedirection implements Redirection {
    INSTANCE;

    public static final String DESC =
        "Lorg/lwjgl/stb/STBImage;stbi_load_from_memory(" +
            "Ljava/nio/ByteBuffer;Ljava/nio/IntBuffer;" +
            "Ljava/nio/IntBuffer;Ljava/nio/IntBuffer;" +
            "I)Ljava/nio/ByteBuffer;";
    private static final BufferedImage DUMMY =
        new BufferedImage(1, 1, BufferedImage.TYPE_INT_RGB);

    @Override
    public Object invoke(Object obj, String desc, Class<?> type, Object... args)
        throws Throwable {
        ByteBuffer buffer = (ByteBuffer) args[0];
        IntBuffer x = (IntBuffer) args[1];
        IntBuffer y = (IntBuffer) args[2];
        IntBuffer channels_in_file = (IntBuffer) args[3];
        int desired_channels = (int) args[4];

        ByteBuffer result = ByteBuffer.wrap(
            new byte[x.get(x.position()) * y.get(y.position())
                * (desired_channels != 0
                ? desired_channels
                : channels_in_file.get(channels_in_file.position()))]);

        BufferedImage image = readImage(buffer);
        x.put(0, image.getWidth());
        y.put(0, image.getHeight());
        // TODO: check discrepancies between desired_channels and actual
        channels_in_file.put(0, desired_channels);
        return result;
    }

    private BufferedImage readImage(ByteBuffer buffer) {
        int position = buffer.position();
        BufferedImage image = null;
        try {
            image = ImageIO.read(new ByteBufferInputStream(buffer));
        } catch (IOException e) {
            e.printStackTrace();
        }

        buffer.position(position);
        return image == null ? DUMMY : image;
    }

}

File: Server-Core/src/main/java/net/rwhps/server/dependent/redirections/lwjgl/STBIImageRedirection.kt
/*
 * Copyright 2020-2023 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package net.rwhps.server.dependent.redirections.lwjgl

import net.rwhps.asm.api.Redirection
import net.rwhps.asm.util.ByteBufferInputStream
import net.rwhps.server.util.alone.annotations.AsmMark
import java.awt.image.BufferedImage
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.IntBuffer
import javax.imageio.ImageIO

@AsmMark.ClassLoaderCompatible
enum class STBIImageRedirection : Redirection {
    INSTANCE;

    @Throws(Throwable::class)
    override fun invoke(obj: Any, desc: String, type: Class<*>?, vararg args: Any): Any {
        val buffer = args[0] as ByteBuffer
        val x = args[1] as IntBuffer
        val y = args[2] as IntBuffer
        val channels_in_file = args[3] as IntBuffer
        val desired_channels = args[4] as Int
        val result = ByteBuffer.wrap(ByteArray(x[x.position()] * y[y.position()] * if (desired_channels != 0) desired_channels else channels_in_file[channels_in_file.position()]))
        val image = readImage(buffer)
        x.put(0, image.width)
        y.put(0, image.height)
        // TODO: check discrepancies between desired_channels and actual
        channels_in_file.put(0, desired_channels)
        return result
    }

    private fun readImage(buffer: ByteBuffer): BufferedImage {
        val position = buffer.position()
        var image: BufferedImage? = null
        try {
            image = ImageIO.read(ByteBufferInputStream(buffer))
        } catch (e: IOException) {
            e.printStackTrace()
        }
        buffer.position(position)
        return image ?: DUMMY
    }

    companion object {
        const val DESC = "Lorg/lwjgl/stb/STBImage;stbi_load_from_memory(Ljava/nio/ByteBuffer;Ljava/nio/IntBuffer;Ljava/nio/IntBuffer;Ljava/nio/IntBuffer;I)Ljava/nio/ByteBuffer;"

        private val DUMMY = BufferedImage(1, 1, BufferedImage.TYPE_INT_RGB)
    }
}